main/src/cgeo/geocaching/calculator/FormulaParser.java

```diff
@@ -0,0 +1,379 @@
+package cgeo.geocaching.calculator;
+
+import cgeo.geocaching.R;
+import cgeo.geocaching.location.Geopoint;
+import cgeo.geocaching.settings.Settings;
+import cgeo.geocaching.utils.CollectionStream;
+import cgeo.geocaching.utils.MatcherWrapper;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Set;
+import java.util.regex.Pattern;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+
+/**
+ * Parse coordinates with formulas.
+ */
+public final class FormulaParser {
+
+    public  static final char WPC_DELIM = '|';
+
+    private static final Pattern PATTERN_BAD_BLANK_COMMA = Pattern.compile("(\\d), (\\d{2,})");
+    private static final Pattern PATTERN_BAD_BLANK_DOT = Pattern.compile("(\\d)\\. (\\d{2,})");
+
+    private static final List<AbstractFormulaParser> parsers = Arrays.asList(new MinDecFormulaParser());
+
+    private Settings.CoordInputFormatEnum desiredFormulaFormat = null;
+
+    public FormulaParser() {
+    }
+
+    public FormulaParser(final Settings.CoordInputFormatEnum formulaFormat) {
+        desiredFormulaFormat = formulaFormat;
+    }
+
+    public static class ParseException extends IllegalArgumentException {
+        private static final long serialVersionUID = 1L;
+        public final int resource;
+
+        public ParseException(final String msg) {
+            super(msg);
+            resource = R.string.err_parse_lat_lon;
+        }
+
+        public ParseException(final String msg, final Geopoint.LatLon faulty) {
+            super(msg);
+            resource = faulty == Geopoint.LatLon.LAT ? R.string.err_parse_lat : R.string.err_parse_lon;
+        }
+    }
+
+    private static class ResultWrapper {
+        private final String result;
+        private final int matcherLength;
+
+        ResultWrapper(final String result, final int stringLength) {
+            this.result = result;
+            this.matcherLength = stringLength;
+        }
+    }
+
+    /**
+     * Abstract parser for coordinate formats.
+     */
+    private abstract static class AbstractFormulaParser {
```

Comment:
I think this abstract class could be replaced by an interface?

Reply:
Probably yes, but with your previous comment, probably that will be done with reusing the other parser...

Reply:
changed to interface